File: src/test/java/com/example/coupon/service/CouponService.java
val couponRepository = mockk<com.example.coupon.dao.CouponRepository>()
    val couponService = com.example.coupon.service.CouponService(couponRepository)

    "redeemCoupon should return true when decrementCouponCount returns a positive value" {
        // Given: A coupon ID and the repository returns 1 (meaning a coupon was decremented)
        val couponId: Long = 1L
        every { couponRepository.decrementCouponCount(couponId) } returns 1

        // When: redeemCoupon is called
        val result: Boolean = couponService.redeemCoupon(couponId)

        // Then: The result should be true
        result shouldBe true
    }

    "redeemCoupon should return false when decrementCouponCount returns 0" {
        // Given: A coupon ID and the repository returns 0 (meaning no coupon was decremented)
        val couponId: Long = 2L
        every { couponRepository.decrementCouponCount(couponId) } returns 0

        // When: redeemCoupon is called
        val result: Boolean = couponService.redeemCoupon(couponId)

        // Then: The result should be false
        result shouldBe false
    }

    "create should save a coupon using the repository" {
        // Given: A CouponCreate object
        val couponCreate = com.example.coupon.controller.CouponCreate("Test Coupon", 10, 5L)
        val expectedCoupon = com.example.coupon.dao.Coupon.builder().name("Test Coupon").count(5L).build()
        every { couponRepository.save(any()) } returns mockk() // We don't care about the return of save.

        // When: create is called
        couponService.create(couponCreate)

        // Then: repository.save should be called
        io.mockk.verify { couponRepository.save(match { it.name == "Test Coupon" && it.count == 5L }) }

    }
})

File: src/test/kotlin/com/example/coupon/service/CouponService.kt
import com.example.coupon.controller.CouponCreate
import com.example.coupon.dao.Coupon
import com.example.coupon.dao.CouponRepository
import com.example.coupon.service.CouponService
import io.kotest.core.spec.style.FunSpec
import io.kotest.matchers.shouldBe
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.springframework.dao.OptimisticLockingFailureException

class CouponServiceTest : FunSpec({

    val couponRepository = mockk<CouponRepository>()
    val couponService = CouponService(couponRepository)

    test("redeemCoupon should decrement coupon count and return true if successful") {
        // Given:
        val couponId: Long = 1
        every { couponRepository.decrementCouponCount(couponId) } returns 1

        // When:
        val result = couponService.redeemCoupon(couponId)

        // Then:
        result shouldBe true
        verify { couponRepository.decrementCouponCount(couponId) }
    }

    test("redeemCoupon should return false if decrementing coupon count fails (count is zero)") {
        // Given:
        val couponId: Long = 1
        every { couponRepository.decrementCouponCount(couponId) } returns 0

        // When:
        val result = couponService.redeemCoupon(couponId)

        // Then:
        result shouldBe false
        verify { couponRepository.decrementCouponCount(couponId) }
    }

    test("create should save a new coupon") {
        // Given:
        val couponCreate = CouponCreate("Test Coupon", 1000, 100L)
        val coupon = Coupon.builder().name(couponCreate.name()).count(couponCreate.count()).build()
        every { couponRepository.save(any<Coupon>()) } returns coupon // Mock the save method

        // When:
        couponService.create(couponCreate)

        // Then:
        verify { couponRepository.save(any<Coupon>()) } // Verify that save was called
    }

})